cpu/stm32_common/periph/flashpage.c

```diff
@@ -15,76 +16,177 @@
-    FLASH->CR &= ~(FLASH_CR_PER);
+    CNTRL_REG &= ~(FLASH_CR_PER);
+
+}
+
+void flashpage_write_raw(void *target_addr, void *data, size_t len)
+{
+    /* The actual minimal block size for writing is 16B, thus we
+     * assert we write on multiples and no less of that length. */
+    assert(!(len % FLASHPAGE_RAW_BLOCKSIZE));
+
+    /* ensure 4 byte aligned writes */
+    assert(!(((unsigned)target_addr % FLASHPAGE_RAW_ALIGNMENT) ||
+            ((unsigned)data % FLASHPAGE_RAW_ALIGNMENT)));
+
+    /* ensure the length doesn't exceed the actual flash size */
+    assert(((unsigned)target_addr + len) <
+           (CPU_FLASH_BASE + (FLASHPAGE_SIZE * FLASHPAGE_NUMOF)) + 1);
+
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    uint32_t *dst = target_addr;
+    uint32_t *data_addr = (uint32_t *)data;
+#else
+    uint16_t *dst = target_addr;
+    uint16_t *data_addr = (uint16_t *)data;
+#endif
+
+    DEBUG("[flashpage_raw] unlocking the flash module\n");
+    _unlock();
+
+    DEBUG("[flashpage] write: now writing the data\n");
+#if !(defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1))
+    /* set PG bit and program page to flash */
+    CNTRL_REG |= FLASH_CR_PG;
+#endif
+    for (size_t i = 0; i < (FLASHPAGE_SIZE / FLASHPAGE_DIV); i++) {
+        DEBUG("[flashpage_raw] writing %c to %p\n", (char)data_addr[i], dst);
+        *dst++ = data_addr[i];
+        while (FLASH->SR & FLASH_SR_BSY) {}
+    }
+
+    /* clear program bit again */
+    CNTRL_REG &= ~(FLASH_CR_PG);
+    DEBUG("[flashpage] write: done writing data\n");
+
+    DEBUG("flashpage_raw] now locking the flash module again\n");
+    _lock();
+}
+
+void flashpage_write(int page, void *data)
+{
+    assert(page < (int)FLASHPAGE_NUMOF);
+
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    /* STM32L0/L1 only supports word sizes */
+    uint32_t *page_addr = flashpage_addr(page);
+#else
+    /* Default is to support half-word sizes */
+    uint16_t *page_addr = flashpage_addr(page);
+#endif
+
+    uint32_t hsi_state = (RCC->CR & RCC_CR_HSION);
+
+    /* the internal RC oscillator (HSI) must be enabled */
+    stmclk_enable_hsi();
+
+    /* unlock the flash module */
+    _unlock();
+
+    /* ERASE sequence */
+    _erase_page(page_addr);
 
     /* WRITE sequence */
     if (data != NULL) {
-        DEBUG("[flashpage] write: now writing the data\n");
-        /* set PG bit and program page to flash */
-        FLASH->CR |= FLASH_CR_PG;
-        for (unsigned i = 0; i < (FLASHPAGE_SIZE / 2); i++) {
-            *page_addr++ = data_addr[i];
-            while (FLASH->SR & FLASH_SR_BSY) {}
-        }
-        /* clear program bit again */
-        FLASH->CR &= ~(FLASH_CR_PG);
-        DEBUG("[flashpage] write: done writing data\n");
+        flashpage_write_raw(page_addr, data, FLASHPAGE_SIZE);
     }
 
     /* finally, lock the flash module again */
-    DEBUG("flashpage] now locking the flash module again\n");
-    FLASH->CR |= FLASH_CR_LOCK;
+    _lock();
```

Comment:
I'm wondering if it wouldn't be cleaner to lock right after erasing, as unlocking/locking are done again in `flashpage_write_raw` in case of writing